src/librustc_mir/borrow_check/nll/region_infer/values.rs

```diff
@@ -202,73 +153,110 @@ impl<N: Idx> RegionValues<N> {
         r: N,
         elem: impl ToElementIndex,
     ) -> bool {
-        let i = self.elements.index(elem);
         debug!("add(r={:?}, elem={:?})", r, elem);
-        self.matrix.add(r, i)
+        elem.add_to_row(self, r)
+    }
+
+    /// Adds all the control-flow points to the values for `r`.
+    crate fn add_all_points(&mut self, r: N) {
+        // FIXME OMG so inefficient. We'll fix later.
+        for p in self.elements.all_points() {
+            self.points.add(r, p);
+        }
     }
 
     /// Add all elements in `r_from` to `r_to` (because e.g. `r_to:
     /// r_from`).
     crate fn add_region(&mut self, r_to: N, r_from: N) -> bool {
-        self.matrix.merge(r_from, r_to)
+        self.points.merge(r_from, r_to) | self.free_regions.merge(r_from, r_to)
+        // FIXME universes?
     }
 
     /// True if the region `r` contains the given element.
-    crate fn contains(&self, r: N, elem: impl ToElementIndex) -> bool {
-        let i = self.elements.index(elem);
-        self.matrix.contains(r, i)
+    crate fn contains(
+        &self,
+        r: N,
+        elem: impl ToElementIndex,
+    ) -> bool {
+        elem.contained_in_row(self, r)
     }
 
-    /// Iterates through each row and the accompanying bit set.
-    pub fn iter_enumerated<'a>(
-        &'a self
-    ) -> impl Iterator<Item = (N, &'a BitVector<RegionElementIndex>)> + 'a {
-        self.matrix.iter_enumerated()
+    /// Iterate through each region that has a value in this set.
+    crate fn regions_with_points<'a>(&'a self) -> impl Iterator<Item = N> {
+        self.points.rows()
     }
 
-    /// Merge a row, `from`, originating in another `RegionValues` into the `into` row.
-    pub fn merge_into(&mut self, into: N, from: &BitVector<RegionElementIndex>) -> bool {
-        self.matrix.merge_into(into, from)
+    /// `self[to] |= values[from]`, essentially: that is, take all the
+    /// elements for the region `from` from `values` and add them to
+    /// the region `to` in `self`.
+    crate fn merge_row<M: Idx>(&mut self, to: N, from: M, values: &RegionValues<M>) {
+        if let Some(set) = values.points.row(from) {
+            self.points.merge_into(to, set);
+        }
+
+        if let Some(set) = values.free_regions.row(from) {
+            self.free_regions.merge_into(to, set);
+        }
     }
 
     /// True if `sup_region` contains all the CFG points that
     /// `sub_region` contains. Ignores universal regions.
-    crate fn contains_points(&self, sup_region: N, sub_region: N) -> bool {
+    crate fn contains_points(
+        &self,
+        sup_region: N,
+        sub_region: N,
+    ) -> bool {
         // This could be done faster by comparing the bitsets. But I
         // am lazy.
```

Comment:
Regarding the comment above the code below *is* now comparing bitsets, right?

Reply:
Er yes should fix :)